### PR TITLE
daemon: Emit Image Create event when image is built

### LINKED ISSUE #47463: No event when building an untagged image
### Description

Building an image with a tag results in `ActionTag` event being sent:

```bash
$ docker events &
$ printf 'FROM alpine\nRUN apk add vim' | docker build -t asdf -
...
2024-02-28T10:37:54.624937133Z image tag sha256:cdd546523a48d2c0932a32faa8c01abc3ebea2724d2571f713175964444d62bd (name=asdf:latest)
 ```

When building an untagged image:
```bash
$ printf 'FROM alpine\nRUN apk add vim' | docker build -
```

the event is not sent, because there's no tagging operation.

This makes it impossible to get a notification via events system when such image is built.

### Reproduce

```bash
$ printf 'FROM alpine\nRUN apk add vim' | docker build -
```

### Expected behavior

Building an untagged image should produce an event

### docker version

```bash
Client:
 Version:           25.0.2
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Thu Feb  1 00:22:01 2024
 OS/Arch:           linux/arm64
 Context:           default

Server:
 Engine:
  Version:          dev
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.7
  Git commit:       b37f8c8070832036d709bcf7d45bce8370b11321
  Built:            Wed Feb 28 10:38:12 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          v1.7.13
  GitCommit:        7c3aca7a610df76212171d200ca3811ff6096eb8
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    25.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/local/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.5
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: dev
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc crun
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7c3aca7a610df76212171d200ca3811ff6096eb8
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
 Kernel Version: 6.6.16-linuxkit
 Operating System: Debian GNU/Linux 12 (bookworm) (containerized)
 OSType: linux
 Architecture: aarch64
 CPUs: 6
 Total Memory: 7.755GiB
 Name: 9300895d92be
 ID: 8f527aca-3d67-48c8-b911-e5e87981ac83
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 28
  Goroutines: 47
  System Time: 2024-02-28T11:06:40.026886584Z
  EventsListeners: 0
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by vvoland**
- closes: https://github.com/moby/moby/issues/47463

**- What I did**
Made `docker build` emit an event `type=image, action=create`, so event subscribers can get a notification when a built image is not tagged.

**- How I did it**

**- How to verify it**
TestBuildEmitsImageCreateEvent

**- Description for the changelog**

```markdown changelog
Emit `image created` events when a new image is created (even if it's not tagged).
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@vvoland looks like test is failing on Windows; probably must be skipped for the "buildkit=true" case;


```
=== FAIL: github.com/docker/docker/integration-cli TestDockerCLIBuildSuite/TestBuildEmitsImageCreateEvent/buildkit=true (2.42s)
    docker_cli_build_test.go:6211: assertion failed: 
        Command:  D:\a\moby\moby\out\docker.exe build --no-cache -
        ExitCode: 1
        Error:    exit status 1
        Stdout:   Sending build context to Docker daemon  2.048kB

        Step 1/2 : FROM busybox
         ---> 052a5e38011c
        Step 2/2 : RUN date >/date
         ---> Running in 2c8181e326e8
        
        Stderr:   The command 'cmd /S /C date >/date' returned a non-zero code: 1
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
```

**Comment by thaJeztah** [comment]
😢 

```
        {"errorDetail":{"message":"failed to read dockerfile: cleaning source path: removing drive letter: UNC paths are not supported"},"error":"failed to read dockerfile: cleaning source path: removing drive letter: UNC paths are not supported"}

**Comment by vvoland** [comment]
I didn't push the buildkit skip on Windows then, should be okay now!

**Comment by vvoland** [comment]
Updated api docs and added an API version gate for the new event.

**Comment by vvoland** [comment]
Removed the version gate, the events will also be emitted on older APIs.

**Comment by thaJeztah** [comment]
> Removed the version gate, the events will also be emitted on older APIs.

should we test it in "integration/" instead? We can have a CLI integration / e2e  test in the CLI repository

**Comment by thaJeztah** [comment]
@vvoland It's a bit of a hack, but looks like setting `DOCKER_API_VERSION=v1.46` allows us to test it in integration-cli with the old CLI 😄 

I opened a quick PR in against your branch; if you merge that, it should appear in this PR (to test it in CI), but please squash with the other commits if you are ok with that approach;

- https://github.com/vvoland/moby/pull/12

**Comment by vvoland** [comment]
> should we test it in "integration/" instead? We can have a CLI integration / e2e test in the CLI repository

There is a test in `integration` too - but it only tests the API path and not the `docker build` (which is wired up a little bit differently).
For the purpose of testing this PR, we can't really move this to the CLI/e2e now.

**Comment by vvoland** [comment]
Looks like enforcing the version via `DOCKER_API_VERSION` doesn't do the trick here 🤔 :
```
 === Failed
=== FAIL: amd64.integration-cli TestDockerCLIEventSuite/TestEventsFilterImageLabels (0.08s)
    docker_cli_events_test.go:360: assertion failed: 3 (int) != 4 (int): Events == [2024-06-17T08:00:06.451183733Z image tag sha256:25840e0118919186301e8d96bef1ca9ff12d1d3cf59dc780754974d2ccacd6a1 (io.docker.testing=image, name=labelfiltertest:latest) 2024-06-17T08:00:06.459109312Z image tag sha256:25840e0118919186301e8d96bef1ca9ff12d1d3cf59dc780754974d2ccacd6a1 (io.docker.testing=image, name=labelfiltertest:tag1) 2024-06-17T08:00:06.466710558Z image tag sha256:25840e0118919186301e8d96bef1ca9ff12d1d3cf59dc780754974d2ccacd6a1 (io.docker.testing=image, name=labelfiltertest:tag2)]
    --- FAIL: TestDockerCLIEventSuite/TestEventsFilterImageLabels (0.08s)

=== FAIL: amd64.integration-cli TestDockerCLIEventSuite/TestEventsFilterType (0.08s)
    docker_cli_events_test.go:587: assertion failed: 3 (int) != 4 (int): Events == [2024-06-17T08:00:08.184129174Z image tag sha256:6f4f4c39bd49e998909d45274af3b4b4912b7d0a9b55633c38237e47976a21fc (io.docker.testing=image, name=labelfiltertest:latest) 2024-06-17T08:00:08.193116473Z image tag sha256:6f4f4c39bd49e998909d45274af3b4b4912b7d0a9b55633c38237e47976a21fc (io.docker.testing=image, name=labelfiltertest:tag1) 2024-06-17T08:00:08.200868189Z image tag sha256:6f4f4c39bd49e998909d45274af3b4b4912b7d0a9b55633c38237e47976a21fc (io.docker.testing=image, name=labelfiltertest:tag2)]
    --- FAIL: TestDockerCLIEventSuite/TestEventsFilterType (0.08s)
```

**Comment by thaJeztah** [comment]
Hm... interesting; was that test changed as part of this PR? Or is it somehow ignoring the version (and sending the events anyway)?

**Comment by vvoland** [comment]
Ah, it's a different test, not the one we added `DOCKER_API_VERSION` to. These tests had adjusted event count to account for the new event, but now that it's version gated it's no longer relevant.

Will fix

File: integration-cli/docker_cli_events_test.go [hunk 1]
```diff
@@ -358,6 +359,10 @@ func (s *DockerCLIEventSuite) TestEventsFilterImageLabels(c *testing.T) {
 	// 2 events from the "docker tag" command, another one is from "docker build"
 	assert.Equal(c, len(events), 3, "Events == %s", events)
 	for _, e := range events {
+		// Ignore image create event emitted by "docker build"
+		if strings.Contains(e, "image create") {
+			continue
+		}
```
Comment (thaJeztah) [inline integration-cli/docker_cli_events_test.go:365]: This one may no longer be needed then as well I think?

File: integration-cli/docker_cli_events_test.go [hunk 2]
```diff
@@ -338,9 +338,10 @@ func (s *DockerCLIEventSuite) TestEventsFilterImageLabels(c *testing.T) {
 	label := "io.docker.testing=image"
 
 	// Build a test image.
-	buildImageSuccessfully(c, name, build.WithDockerfile(fmt.Sprintf(`
-		FROM busybox:latest
-		LABEL %s`, label)))
+	buildImageSuccessfully(c, name,
+		build.WithDockerfile("FROM busybox:latest\nLABEL "+label),
+		build.WithoutCache, // Make sure image is actually built (and we always get the image create event to get consistent event count)
```
Comment (thaJeztah) [inline integration-cli/docker_cli_events_test.go:343]: And this comment mentions "image create event", so may need to be adjusted.

Adding the `WithoutCache` is still a good improvement though.

File: integration-cli/docker_cli_events_test.go [hunk 3]
```diff
@@ -560,9 +565,10 @@ func (s *DockerCLIEventSuite) TestEventsFilterType(c *testing.T) {
...
```
Comment (thaJeztah) [inline integration-cli/docker_cli_events_test.go:570]: Same here

File: integration-cli/docker_cli_events_test.go [hunk 4]
```diff
@@ -580,6 +586,10 @@ func (s *DockerCLIEventSuite) TestEventsFilterType(c *testing.T) {
 	// 2 events from the "docker tag" command, another one is from "docker build"
 	assert.Equal(c, len(events), 3, "Events == %s", events)
 	for _, e := range events {
+		// Ignore image create event emitted by "docker build".
+		if strings.Contains(e, "image create") {
+			continue
+		}
```
Comment (thaJeztah) [inline integration-cli/docker_cli_events_test.go:592]: And here 😅

File: integration-cli/docker_cli_events_test.go (hunk 1 above)
Comment (vvoland) [reply integration-cli/docker_cli_events_test.go]: Good catch, removed

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

File: builder/builder-next/exporter/wrapper.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+	"github.com/docker/docker/builder/builder-next/exporter/overrides"
+	"github.com/moby/buildkit/exporter"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type ImageExportedByBuildkit interface {
+	ImageExportedByBuildkit(ctx context.Context, id string, desc ocispec.Descriptor) error
```
Comment (laurazard) [inline builder/builder-next/exporter/wrapper.go:15]: Any benefits of defining `ImageExportedByBuildkit` as an interface and having the daemon implement it over a more explicit approach such as 
```
type BuildkitExportCallback func(...) error
```

and having the daemon pass in a function?

Comment (vvoland) [reply builder/builder-next/exporter/wrapper.go]: In this case I don't think there's any benefit. Changed to your suggestion as it's a bit less mouthful.

**Review by laurazard — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
still LGTM

**Pull request merged**